scripts/mbedtls_framework/code_wrapper/psa_buffer.py

```diff
@@ -14,13 +15,42 @@ class BufferParameter:
     def __init__(self, i: int, is_output: bool,
                  buffer_name: str, size_name: str) -> None:
         """Initialize the parameter information.
+
         i is the index of the function argument that is the pointer to the buffer.
         The size is argument i+1. For a variable-size output, the actual length
         goes in argument i+2.
+
         buffer_name and size_names are the names of arguments i and i+1.
         This class does not yet help with the output length.
         """
         self.index = i
         self.buffer_name = buffer_name
         self.size_name = size_name
         self.is_output = is_output
+
+    def poison_write(self,
+                     out: typing_util.Writable,
+                     poison: bool) -> None:
+        """Write poisoning or unpoisoning code for a buffer parameter.
+
+        Write poisoning code if poison is true, unpoisoning code otherwise.
+        """
+        out.write('    MBEDTLS_TEST_MEMORY_{}({}, {});\n'.format(
+            'POISON' if poison else 'UNPOISON',
+            self.buffer_name, self.size_name
+        ))
+
+    @staticmethod
+    def poison_multi_write(out: typing_util.Writable,
```

Comment:
This is tricky. 

The point of a data class is that it has knowlege of the data that it contains, and should be able to manipulate or present it in a custom format, without having to expose them to the calling one.

Otherwise we end up splitting the class's data manipulation accross two classes which is not ideal

I will propose the following:

1. Keep the text formatting as class function.
2. Set the indentation as an input parameter
3. Move the out.write() logic to the PSA_Wrapper.

e.g 

      def poison_wrap(self, poison: bool, ident_lv = "1") -> None:
          """Return a POISON/UNPOISON pr.
  
          return "{}MBEDTLS_TEST_MEMORY_{}({}, {});\n".format((ident_lv * 4) * " ",
                                                              POISON' if poison else 'UNPOISON',
                                                              self.buffer_name, self.size_name),

Reply:
> The point of a data class is that it has knowlege of the data that it contains, and should be able to manipulate or present it in a custom format, without having to expose them to the calling one.

Not necessarily. The `list` class doesn't have any knowledge of how to present the data that it contains. Some classes are collections whose job is to hold data however it's told; such classes allow the data to be accessed freely. Other classes have a job that includes protecting the integrity of the data; such classes only present formatted views of the data and only allow selected modifications. This is not a binary classification: many classes fall somewhere on the spectrum, possibly at different places for different parts of the data.

In this case, I prefer the way I did it originally. Keep `BufferParameter` as a data-oriented class that knows how to store the data, but not how to process it. This follows the general software engineering principle of keeping concerns separate (also known as single-responsibility principle and various other names). More concretely, here, it means that the wrapper generator classes keep full control as to how to format the C code, so maintainers don't have to worry about how a completely separate class formats C code. Conversely, `BufferParameter` might be used in contexts other than generating C code (for example to generate test cases in some systematic way), and in those contexts knowledge of C code formatting would be irrelevant, which shows that it isn't the job of `BufferParameter` to worry about C code formatting.

Reply:
Also why shouldn't multi-write be a utliity method of the module instead of a statit method?

It needs to do 3 things:

- Check if the iterable parameters has any entries
- If it has add the `#if !defined(MBEDTLS_PSA_ASSUME_EXCLUSIVE_BUFFERS` guards
- If it has parameters, iterate through them and write them.

Moving the logic back to the psa_wrapper without that utility,  looks like that:


        if buffer_parameters:
            out.write('#if !defined(MBEDTLS_PSA_ASSUME_EXCLUSIVE_BUFFERS)\n')
            for param in buffer_parameters:
                out.write(param.poison_wrap(True))
            out.write('#endif /* !defined(MBEDTLS_PSA_ASSUME_EXCLUSIVE_BUFFERS) */\n')

        super()._write_function_call(out, function, argument_names)

        if buffer_parameters:
            out.write('#if !defined(MBEDTLS_PSA_ASSUME_EXCLUSIVE_BUFFERS)\n')
            for param in buffer_parameters:
                out.write(param.poison_wrap(False))
            out.write('#endif /* !defined(MBEDTLS_PSA_ASSUME_EXCLUSIVE_BUFFERS) */\n')

Which looks unnessary more redudant compared to:

        BufferParameter.poison_multi_write(out, buffer_parameters, True)
        super()._write_function_call(out, function, argument_names)
        BufferParameter.poison_multi_write(out, buffer_parameters, False)